File: typespec/employer/applications.go
package employer

import (
	"time"

	"github.com/psankar/vetchi/typespec/common"
)

type ApplicationColorTag string

const (
	// Any change here should reflect in the IsValid() method too
	GreenApplicationColorTag  ApplicationColorTag = "GREEN"
	YellowApplicationColorTag ApplicationColorTag = "YELLOW"
	RedApplicationColorTag    ApplicationColorTag = "RED"
)

func (c ApplicationColorTag) IsValid() bool {
	return c == GreenApplicationColorTag ||
		c == YellowApplicationColorTag ||
		c == RedApplicationColorTag
}

type GetApplicationsRequest struct {
	State          common.ApplicationState `json:"state"            validate:"validate_application_state"`
	SearchQuery    *string                 `json:"search_query"     validate:"omitempty,max=25"`
	ColorTagFilter *ApplicationColorTag    `json:"color_tag_filter" validate:"omitempty"`
	OpeningID      string                  `json:"opening_id"       validate:"required"`
	PaginationKey  *string                 `json:"pagination_key"   validate:"omitempty"`
	Limit          int64                   `json:"limit"            validate:"required,min=0,max=40"`
}

type Endorser struct {
	FullName              string   `json:"full_name"`
	ShortBio              string   `json:"short_bio"`
	Handle                string   `json:"handle"`
	CurrentCompanyDomains []string `json:"current_company_domains"`
}

type Application struct {
	ID                        string                  `json:"id"`
	CoverLetter               *string                 `json:"cover_letter,omitempty"`
	CreatedAt                 time.Time               `json:"created_at"`
	HubUserHandle             string                  `json:"hub_user_handle"`
	HubUserName               string                  `json:"hub_user_name"`
	HubUserLastEmployerDomain *string                 `json:"hub_user_last_employer_domain,omitempty"`
	State                     common.ApplicationState `json:"state"`
	ColorTag                  *ApplicationColorTag    `json:"color_tag,omitempty"`
	Endorsers                 []Endorser              `json:"endorsers"`
}

type SetApplicationColorTagRequest struct {
	ApplicationID string              `json:"application_id" validate:"required"`
	ColorTag      ApplicationColorTag `json:"color_tag"      validate:"required,validate_application_color_tag"`
}

type RemoveApplicationColorTagRequest struct {
	ApplicationID string `json:"application_id" validate:"required"`
}

type ShortlistApplicationRequest struct {
	ApplicationID string `json:"application_id" validate:"required"`
}

type RejectApplicationRequest struct {
	ApplicationID string `json:"application_id" validate:"required"`
}

type GetResumeRequest struct {
	ApplicationID string `json:"application_id" validate:"required"`
}

File: typespec/employer/applications.ts
import { ApplicationState } from "../common/applications";

export type ApplicationColorTag = "GREEN" | "YELLOW" | "RED";

export const ApplicationColorTags = {
  GREEN: "GREEN" as ApplicationColorTag,
  YELLOW: "YELLOW" as ApplicationColorTag,
  RED: "RED" as ApplicationColorTag,
} as const;

export interface GetApplicationsRequest {
  state: ApplicationState;
  search_query?: string;
  color_tag_filter?: ApplicationColorTag;
  opening_id: string;
  pagination_key?: string;
  limit: number;
}

export interface Endorser {
  full_name: string;
  short_bio: string;
  handle: string;
  current_company_domains?: string[];
}

export interface ModelScore {
  model_name: string;
  score: number;
}

export interface Application {
  id: string;
  cover_letter?: string;
  created_at: Date;
  hub_user_handle: string;
  hub_user_name: string;
  hub_user_short_bio: string;
  hub_user_last_employer_domains?: string[];
  state: ApplicationState;
  color_tag?: ApplicationColorTag;
  endorsers: Endorser[];
  scores: ModelScore[];
}

export interface SetApplicationColorTagRequest {
  application_id: string;
  color_tag: ApplicationColorTag;
}

export interface RemoveApplicationColorTagRequest {
  application_id: string;
}

export interface ShortlistApplicationRequest {
  application_id: string;
}

export interface RejectApplicationRequest {
  application_id: string;
}

export interface GetResumeRequest {
  application_id: string;
}

export function isValidApplicationColorTag(
  tag: string
): tag is ApplicationColorTag {
  return Object.values(ApplicationColorTags).includes(
    tag as ApplicationColorTag
  );
}
